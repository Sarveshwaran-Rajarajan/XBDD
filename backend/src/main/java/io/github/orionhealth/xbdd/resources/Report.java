package io.github.orionhealth.xbdd.resources;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;

import javax.ws.rs.BeanParam;
import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
import org.glassfish.jersey.media.multipart.FormDataParam;
import org.springframework.beans.factory.annotation.Autowired;

import io.github.orionhealth.xbdd.mappers.FeatureMapper;
import io.github.orionhealth.xbdd.model.common.Summary;
import io.github.orionhealth.xbdd.model.common.Tag;
import io.github.orionhealth.xbdd.model.common.TestingTips;
import io.github.orionhealth.xbdd.model.junit.JUnitFeature;
import io.github.orionhealth.xbdd.model.xbdd.XbddFeature;
import io.github.orionhealth.xbdd.model.xbdd.XbddFeatureSummary;
import io.github.orionhealth.xbdd.model.xbdd.XbddScenario;
import io.github.orionhealth.xbdd.persistence.FeatureDao;
import io.github.orionhealth.xbdd.persistence.ImageDao;
import io.github.orionhealth.xbdd.persistence.StatsDao;
import io.github.orionhealth.xbdd.persistence.SummaryDao;
import io.github.orionhealth.xbdd.persistence.TestingTipsDao;
import io.github.orionhealth.xbdd.persistence.UsersDao;
import io.github.orionhealth.xbdd.util.Coordinates;
import io.github.orionhealth.xbdd.util.LoggedInUserUtil;
import io.github.orionhealth.xbdd.util.TestingTipUtil;

@Path("/reports")
public class Report {

	@Autowired
	private FeatureDao featureDao;

	@Autowired
	private ImageDao imageDao;

	@Autowired
	private SummaryDao summaryDao;

	@Autowired
	private StatsDao statsDao;

	@Autowired
	private UsersDao usersDao;

	@Autowired
	private TestingTipsDao testingTipsDao;

	@GET
	@Path("/{product}/{major}.{minor}.{servicePack}/{build}")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getReportByProductVersionId(@BeanParam final Coordinates coordinates) {
		final List<XbddFeature> features = this.featureDao.getFeatures(coordinates);
		final Map<String, TestingTips> testingTips = this.testingTipsDao.getLatestTestingTips(coordinates);

		for (final XbddFeature feature : features) {
			if (feature.getElements() != null) {
				for (final XbddScenario scenario : feature.getElements()) {
					final String tipKey = TestingTipUtil.getMapKey(feature, scenario);
					if (testingTips.containsKey(tipKey)) {
						scenario.setTestingTips(testingTips.get(tipKey).getTestingTips());
					}
				}
			}
		}

		return Response.ok(features).build();
	}

	@GET
	@Path("/summary")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getSummaryOfAllReports() {
		// TODO this doesn't actually do anything atm as we dont have users but this is how you would use it
		final List<String> favourites = this.usersDao.getUserFavourites(LoggedInUserUtil.getLoggedInUser().getDisplay());
		final List<Summary> summaries = this.summaryDao.getSummaries();

		summaries.forEach(summary -> summary.setFavourite(favourites.contains(summary.getCoordinates().getProduct())));

		return Response.ok(summaries).build();
	}

	@GET
	@Produces(MediaType.APPLICATION_JSON)
	@Path("/featureIndex/{product}/{major}.{minor}.{servicePack}/{build}")
	public Response getFeatureIndexForReport(@BeanParam final Coordinates coordinates) {
		final List<XbddFeatureSummary> featureSummaries = this.featureDao.getFeatureSummaries(coordinates);
		return Response.ok(featureSummaries).build();
	}

	@GET
	@Produces(MediaType.APPLICATION_JSON)
	@Path("/tags/{product}/{major}.{minor}.{servicePack}/{build}")
	public Response getTagList(@BeanParam final Coordinates coordinates) {
		final List<XbddFeature> features = this.featureDao.getFeatures(coordinates);

		final Set<Tag> tags = new TreeSet<>(Comparator.comparing(Tag::getName));

		for (final XbddFeature feature : features) {
			if (feature.getTags() != null) {
				tags.addAll(feature.getTags());
			}

			if (feature.getElements() != null) {
				for (final XbddScenario scenario : feature.getElements()) {
					if (scenario.getTags() != null) {
						tags.addAll(scenario.getTags());
					}
				}
			}
		}

		return Response.ok(tags).build();
	}

	@PUT
	@Path("/{product}/{major}.{minor}.{servicePack}/{build}")
	@Produces(MediaType.APPLICATION_JSON)
	@Consumes(MediaType.APPLICATION_JSON)
	public Response putReport(@BeanParam final Coordinates coordinates, final List<JUnitFeature> root) {
		final FeatureMapper featureMapper = new FeatureMapper(this.imageDao);
		this.summaryDao.updateSummary(coordinates);

		final List<XbddFeature> mappedFeatures = root.stream().map(feature -> featureMapper.map(feature, coordinates))
				.collect(Collectors.toList());
		this.featureDao.saveFeatures(mappedFeatures);

		final List<XbddFeature> persistedFeatures = this.featureDao.getFeatures(coordinates);

		this.statsDao.updateStatsForFeatures(coordinates, persistedFeatures);

		return Response.ok(persistedFeatures).build();
	}

	@POST
	@Path("/{product}/{major}.{minor}.{servicePack}/{build}")
	@Consumes(MediaType.MULTIPART_FORM_DATA)
	public Response uploadFile(
			@BeanParam final Coordinates coord,
			@FormDataParam("file") final List<JUnitFeature> root,
			@FormDataParam("file") final FormDataContentDisposition fileDetail) {
		putReport(coord, root);
		return Response.ok().build();
	}
}
