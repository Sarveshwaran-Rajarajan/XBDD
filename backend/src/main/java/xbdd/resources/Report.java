/**
 * Copyright (C) 2015 Orion Health (Orchestral Development Ltd)
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package xbdd.resources;

import com.mongodb.*;
import com.mongodb.client.MongoDatabase;
import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
import org.glassfish.jersey.media.multipart.FormDataParam;
import xbdd.factory.MongoDBAccessor;
import xbdd.mappers.FeatureMapper;
import xbdd.model.junit.JUnitFeature;
import xbdd.model.xbdd.XbddFeature;
import xbdd.persistence.FeatureDao;
import xbdd.persistence.ImageDao;
import xbdd.persistence.StatsDao;
import xbdd.persistence.SummaryDao;
import xbdd.util.Coordinates;
import xbdd.util.SerializerUtil;

import javax.inject.Inject;
import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Path("/reports")
public class Report {

	private final MongoDBAccessor client;
	private final FeatureDao featureDao;
	private final ImageDao imageDao;
	private final SummaryDao summaryDao;
	private final StatsDao statsDao;

	@Inject
	public Report(final MongoDBAccessor client) {
		this.client = client;
		this.featureDao = new FeatureDao(client);
		this.imageDao = new ImageDao(client);
		this.summaryDao = new SummaryDao(client);
		this.statsDao = new StatsDao(client);
	}

	@GET
	@Path("/{product}/{major}.{minor}.{servicePack}/{build}")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getReportByProductVersionId(@BeanParam final Coordinates coordinates,
			@QueryParam("searchText") final String searchText, @QueryParam("viewPassed") final Integer viewPassed,
			@QueryParam("viewFailed") final Integer viewFailed,
			@QueryParam("viewUndefined") final Integer viewUndefined, @QueryParam("viewSkipped") final Integer viewSkipped,
			@QueryParam("start") final String start, @QueryParam("limit") final Integer limit) {

		final BasicDBObject example = xbdd.resources.QueryBuilder.getInstance().buildFilterQuery(coordinates, searchText, viewPassed,
				viewFailed, viewUndefined, viewSkipped, start);

		final DB db = this.client.getDB("bdd");
		final DBCollection collection = db.getCollection("features");
		final DBCursor cursor = collection.find(example).sort(Coordinates.getFeatureSortingObject());
		try {
			if (limit != null) {
				cursor.limit(limit);
			}
			final BasicDBList featuresToReturn = new BasicDBList();
			while (cursor.hasNext()) {
				featuresToReturn.add(cursor.next());
			}
			embedTestingTips(featuresToReturn, coordinates, db);
			return Response.ok(SerializerUtil.serialise(featuresToReturn)).build();
		} finally {
			cursor.close();
		}
	}

	@GET
	@Path("/summary")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getSummaryOfAllReports(@Context final HttpServletRequest req) {
		final DB db = this.client.getDB("bdd");
		final DBCollection collection = db.getCollection("summary");
		final DBCollection usersCollection = db.getCollection("users");

		final BasicDBObject user = new BasicDBObject();
		user.put("user_id", req.getRemoteUser());

		final DBCursor userCursor = usersCollection.find(user);
		DBObject userFavourites;

		if (userCursor.count() != 1) {
			userFavourites = new BasicDBObject();
		} else {
			final DBObject uDoc = userCursor.next();
			if (uDoc.containsField("favourites")) {
				userFavourites = (DBObject) uDoc.get("favourites");
			} else {
				userFavourites = new BasicDBObject();
			}
		}

		final DBCursor cursor = collection.find();

		try {
			final BasicDBList returns = new BasicDBList();
			DBObject doc;

			while (cursor.hasNext()) {
				doc = cursor.next();
				final String product = ((String) ((DBObject) doc.get("coordinates")).get("product"));
				if (userFavourites.containsField(product)) {
					doc.put("favourite", userFavourites.get(product));
				} else {
					doc.put("favourite", false);
				}
				returns.add(doc);
			}

			return Response.ok(SerializerUtil.serialise(returns)).build();
		} finally {
			cursor.close();
		}
	}

	@GET
	@Produces(MediaType.APPLICATION_JSON)
	@Path("/featureIndex/{product}/{major}.{minor}.{servicePack}/{build}")
	public Response getFeatureIndexForReport(@BeanParam final Coordinates coordinates,
			@QueryParam("searchText") final String searchText, @QueryParam("viewPassed") final Integer viewPassed,
			@QueryParam("viewFailed") final Integer viewFailed,
			@QueryParam("viewUndefined") final Integer viewUndefined, @QueryParam("viewSkipped") final Integer viewSkipped,
			@QueryParam("start") final String start) {

		final BasicDBObject example = QueryBuilder.getInstance().buildFilterQuery(coordinates, searchText, viewPassed,
				viewFailed, viewUndefined, viewSkipped, start);
		final DB db = this.client.getDB("bdd");
		final DBCollection featuresCollection = db.getCollection("features");
		final DBCursor features = featuresCollection.find(example,
				new BasicDBObject("id", 1).append("name", 1).append("calculatedStatus", 1)
						.append("originalAutomatedStatus", 1).append("tags", 1).append("uri", 1))
				.sort(Coordinates.getFeatureSortingObject());
		final BasicDBList featureIndex = new BasicDBList();
		try {
			for (final Object o : features) {
				featureIndex.add(o);
			}
		} finally {
			features.close();
		}
		return Response.ok(SerializerUtil.serialise(featureIndex)).build();
	}

	protected void embedTestingTips(final BasicDBList featureList, final Coordinates coordinates, final DB db) {
		for (final Object feature : featureList) {
			Feature.embedTestingTips((DBObject) feature, coordinates, db);
		}
	}

	@GET
	@Produces(MediaType.APPLICATION_JSON)
	@Path("/tags/{product}/{major}.{minor}.{servicePack}/{build}")
	public Response getTagList(@BeanParam final Coordinates coordinates) {
		final DB bdd = this.client.getDB("bdd");
		final DBCollection features = bdd.getCollection("features");
		List<BasicDBObject> objectList = new ArrayList<>();
		// Build objects for aggregation pipeline
		// id option: returns each tag with a list of associated feature ids
		objectList.add(new BasicDBObject("$match", coordinates.getReportCoordinatesQueryObject()));
		final DBObject fields = new BasicDBObject("tags.name", 1);
		fields.put("_id", 0); // comment out for id option
		objectList.add(new BasicDBObject("$project", fields));
		objectList.add(new BasicDBObject("$unwind", "$tags"));
		final DBObject groupFields = new BasicDBObject("_id", "$tags.name");
		// groupFields.put("features", new BasicDBObject("$addToSet", "$_id")); //comment in for id option
		groupFields.put("amount", new BasicDBObject("$sum", 1));
		objectList.add(new BasicDBObject("$group", groupFields));
		objectList.add(new BasicDBObject("$sort", new BasicDBObject("amount", -1)));

		AggregationOptions options = AggregationOptions.builder().build();

		final Cursor output = features.aggregate(objectList, options);

		// get _ids from each entry of output.result
		final BasicDBList returns = new BasicDBList();
		while (output.hasNext()) {
			returns.add(output.next().get("_id").toString());
		}
		return Response.ok(SerializerUtil.serialise(returns)).build();
	}

	@PUT
	@Path("/{product}/{major}.{minor}.{servicePack}/{build}")
	@Produces(MediaType.APPLICATION_JSON)
	@Consumes(MediaType.APPLICATION_JSON)
	public Response putReport(@BeanParam final Coordinates coordinates, final List<JUnitFeature> root) {
		final MongoDatabase bdd = this.client.getDatabase("bdd");

		final FeatureMapper featureMapper = new FeatureMapper(this.imageDao);
		this.summaryDao.updateSummary(coordinates);

		List<XbddFeature> mappedFeatures = root.stream().map(feature -> featureMapper.map(feature, coordinates)).collect(Collectors.toList());
		this.featureDao.saveFeatures(mappedFeatures);

		List<XbddFeature> persistedFeatures = this.featureDao.getFeatures(coordinates);

		this.statsDao.updateStatsForFeatures(coordinates, persistedFeatures);

		return Response.ok(SerializerUtil.serialise(persistedFeatures)).build();
	}

	@POST
	@Path("/{product}/{major}.{minor}.{servicePack}/{build}")
	@Consumes(MediaType.MULTIPART_FORM_DATA)
	public Response uploadFile(
			@BeanParam final Coordinates coord,
			@FormDataParam("file") final List<JUnitFeature> root,
			@FormDataParam("file") final FormDataContentDisposition fileDetail) {
		putReport(coord, root);
		return Response.ok().build();

	}
}
