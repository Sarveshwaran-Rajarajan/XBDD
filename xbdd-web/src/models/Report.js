import Tag from './Tag';

class Report {
    constructor(data) {
        this.tagList = data.map(item => new Tag(item));
    }
}

export default Report;
