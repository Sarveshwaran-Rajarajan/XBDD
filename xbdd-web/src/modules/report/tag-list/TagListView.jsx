import React from "react";
import { PropTypes } from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import Tag from "../../../models/Tag";
import TagListItemView from "./widgets/TagListItemView";
import TagListFilterButtonsView from "./widgets/TagListFilterButtonsView";
import { tagListStyles } from "./styles/TagListStyles";
import { Card } from "@material-ui/core";

const mapTagToTagListItem = (tag, itemsUIState, filterStates, onSelectTag) => (
  <TagListItemView tag={tag} key={tag.name} itemsUIState={itemsUIState} filterStates={filterStates} onSelectTag={onSelectTag} />
);

const renderList = (className, tags, itemsUIState, filterStates, onSelectTag) => {
  if (!tags.length) {
    return null;
  }
  return (
    <Card raised className={className}>
      <List component="ul">{tags.map(tag => mapTagToTagListItem(tag, itemsUIState[tag.name], filterStates, onSelectTag))}</List>
    </Card>
  );
};

const TagListView = props => (
  <div className={props.classes.xbddTagListContainer}>
    <TagListFilterButtonsView filterStates={props.filterStates} onFilterButtonClick={props.onFilterButtonClick} />
    {renderList(props.classes.xbddTagList, props.tags, props.itemsUIState, props.filterStates, props.onSelectTag)}
  </div>
);

TagListView.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.instanceOf(Tag)),
  filterStates: PropTypes.shape({}).isRequired,
  onSelectTag: PropTypes.func,
  onFilterButtonClick: PropTypes.func,
  classes: PropTypes.shape({
    xbddTagListContainer: PropTypes.string,
    xbddTagList: PropTypes.string,
  }).isRequired,
};

TagListView.defaultProps = {
  tags: [],
  selectedTag: null,
  onSelectTag: () => {},
  onFilterButtonClick: () => {},
};

export default withStyles(tagListStyles)(TagListView);
