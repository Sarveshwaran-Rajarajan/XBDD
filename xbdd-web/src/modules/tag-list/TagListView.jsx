import React from "react";
import { PropTypes } from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import Tag from "../../models/Tag";
import TagListItemView from "./widgets/TagListItemView";
import TagListFilterButtonsView from "./widgets/TagListFilterButtonsView";
import { tagListStyles } from "./styles/TagListStyles";

const mapTagToTagListItem = (tag, selectedTag, onSelectTag) => (
    <TagListItemView
        tag={tag}
        key={tag.name}
        onSelectTag={onSelectTag}
        isSelected={tag === selectedTag}
    />
);

const renderList = (tags, selectedTag, onSelectTag) => (
    <List component="ul">
        {tags.map(tag => mapTagToTagListItem(tag, selectedTag, onSelectTag))}
    </List>
);

const TagListView = props => (
    <div className={props.classes.xbddTagListContainer}>
        <TagListFilterButtonsView
            filterStates={props.filterStates}
            onFilterButtonClick={props.onFilterButtonClick}
        />
        <div className={props.classes.xbddTagList}>
            {renderList(props.tags, props.selectedTag, props.onSelectTag)}
        </div>
    </div>
);

TagListView.propTypes = {
    tags: PropTypes.arrayOf(PropTypes.instanceOf(Tag)),
    selectedTag: PropTypes.instanceOf(Tag),
    filterStates: PropTypes.shape({
        passedSelected: PropTypes.bool,
        undefinedSelected: PropTypes.bool,
        failedSelected: PropTypes.bool,
        skippedSelected: PropTypes.bool,
    }).isRequired,
    onSelectTag: PropTypes.func,
    onFilterButtonClick: PropTypes.func,
    classes: PropTypes.object.isRequired,
};

TagListView.defaultProps = {
    tags: [],
    selectedTag: null,
    onSelectTag: () => {},
    onFilterButtonClick: () => {},
};

export default withStyles(tagListStyles)(TagListView);
