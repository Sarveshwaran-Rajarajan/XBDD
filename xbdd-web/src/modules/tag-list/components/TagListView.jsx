import React from 'react';
import { PropTypes } from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import Tag from '../../../models/Tag';
import TagListItemView from './TagListItemView';
import TagListFilterButtonsView from './TagListFilterButtonsView';
import { tagListStyles } from '../styles/TagListStyles';

const mapTagToTagListItem = (tag, selectedTag, onSelectTag) => (
    <TagListItemView
        tag={tag}
        key={tag.name}
        onSelectTag={onSelectTag}
        isSelected={tag === selectedTag}
    />);

const renderList = (tags, selectedTag, onSelectTag) => (
    <List component="ul" >
        {tags.map(tag => mapTagToTagListItem(tag, selectedTag, onSelectTag))}
    </List>);

const TagListView = props => (
    <div className={props.classes.xbddTagListContainer}>
        <TagListFilterButtonsView
            filterStates={props.filterStates}
            onFilterButtonClick={props.onFilterButtonClick}
        />
        <div className={props.classes.xbddTagList}>
            {renderList(props.tags, props.selectedTag, props.onSelectTag)}
        </div>
    </div>
);

TagListView.propTypes = {
    tags: PropTypes.arrayOf(PropTypes.instanceOf(Tag)).isRequired,
    selectedTag: PropTypes.instanceOf(Tag),
    filterStates: PropTypes.object.isRequired,
    onSelectTag: PropTypes.func.isRequired,
    onFilterButtonClick: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired,
};

TagListView.defaultProps = {
    selectedTag: null,
};

export default withStyles(tagListStyles)(TagListView);
