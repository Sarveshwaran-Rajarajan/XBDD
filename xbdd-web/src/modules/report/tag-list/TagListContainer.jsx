import React, { Component } from "react";
import { PropTypes } from "prop-types";
import Report from "../../../models/Report";
import TagListView from "./TagListView";

class TagListContainer extends Component {
  constructor(props) {
    super(props);

    const itemsUIState = {};
    props.report.tagList.forEach(tag => {
      itemsUIState[tag.name] = {
        expanded: false,
      };
    });

    this.state = {
      filterStates: {
        passedSelected: true,
        undefinedSelected: true,
        failedSelected: true,
        skippedSelected: true,
      },
      itemsUIState: itemsUIState,
    };
    this.onFilterButtonClick = this.onFilterButtonClick.bind(this);
    this.onSelectTag = this.onSelectTag.bind(this);
  }

  componentDidCatch(error, info) {
    console.error(error, info);
  }

  onFilterButtonClick(stateAttribute) {
    this.setState(prevState =>
      Object.assign(prevState.filterStates, {
        [stateAttribute]: !prevState.filterStates[stateAttribute],
      }));
  }

  onSelectTag(tag) {
    this.setState(prevState => {
      const newState = Object.assign({}, prevState);
      const storedItem = newState.itemsUIState[tag.name];
      storedItem.expanded = !storedItem.expanded;
      return newState;
    });
  }

  filterTags() {
    const tags = this.props.report.tagList;

    return tags.filter(
      tag =>
        (this.state.filterStates.passedSelected && tag.containsPassed) ||
        (this.state.filterStates.failedSelected && tag.containsFailed) ||
        (this.state.filterStates.undefinedSelected && tag.containsUndefined) ||
        (this.state.filterStates.skippedSelected && tag.containsSkipped)
    );
  }

  render() {
    return (
      <TagListView
        tags={this.filterTags()}
        filterStates={this.state.filterStates}
        itemsUIState={this.state.itemsUIState}
        onSelectTag={this.onSelectTag}
        onFilterButtonClick={this.onFilterButtonClick}
      />
    );
  }
}

TagListContainer.propTypes = {
  report: PropTypes.instanceOf(Report).isRequired,
};

export default TagListContainer;
