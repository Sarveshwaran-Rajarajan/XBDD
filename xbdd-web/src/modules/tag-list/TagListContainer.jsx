import React, { Component } from "react";
import { PropTypes } from "prop-types";
import Report from "../../models/Report";
import TagListView from "./TagListView";

class TagListContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      filterStates: {
        passedSelected: true,
        undefinedSelected: true,
        failedSelected: true,
        skippedSelected: true,
      },
      selectedTag: null,
    };
    this.onFilterButtonClick = this.onFilterButtonClick.bind(this);
    this.onSelectTag = this.onSelectTag.bind(this);
  }

  componentDidCatch(error, info) {
    console.error(error, info);
  }

  onFilterButtonClick(stateAttribute) {
    this.setState(prevState =>
      Object.assign(prevState.filterStates, {
        [stateAttribute]: !prevState.filterStates[stateAttribute],
      }));
  }

  onSelectTag(tag) {
    this.setState({
      selectedTag: tag,
    });
  }

  filterTags() {
    const tags = this.props.report.tagList;

    return tags.filter(
      tag =>
        (this.state.filterStates.passedSelected && tag.containsPassed) ||
        (this.state.filterStates.undefinedSelected && tag.containsUndefined) ||
        (this.state.filterStates.failedSelected && tag.containsFailed) ||
        (this.state.filterStates.skippedSelected && tag.containsSkipped)
    );
  }

  render() {
    return (
      <TagListView
        tags={this.filterTags()}
        selectedTag={this.state.selectedTag}
        filterStates={this.state.filterStates}
        onSelectTag={this.onSelectTag}
        onFilterButtonClick={this.onFilterButtonClick}
      />
    );
  }
}

TagListContainer.propTypes = {
  report: PropTypes.instanceOf(Report).isRequired,
};

export default TagListContainer;
