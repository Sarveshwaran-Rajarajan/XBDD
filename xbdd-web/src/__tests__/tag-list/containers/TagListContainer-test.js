import React from "react";
import { shallow } from "enzyme";
import Report from "../../../models/Report";
import TagListContainer from "../../../modules/tag-list/TagListContainer";
import TagListView from "../../../modules/tag-list/TagListView";

const dummyReport = new Report([
  {
    tag: "testTagOne",
    features: [{ elements: [{ steps: [{ result: { status: "passed" } }] }] }],
  },
]);

describe("TagListContainer", () => {
  test("renders", () => {
    const tagList = shallow(<TagListContainer report={dummyReport} />);
    expect(tagList).toMatchSnapshot();
  });

  test("renders the TagListView", () => {
    const tagList = shallow(<TagListContainer report={dummyReport} />);
    expect(tagList.find(TagListView)).toHaveLength(1);
  });
});
