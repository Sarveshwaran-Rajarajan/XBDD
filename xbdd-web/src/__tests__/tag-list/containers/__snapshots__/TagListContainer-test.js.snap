// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagListContainer renders 1`] = `ShallowWrapper {}`;
